cs169/BJC-Teacher-Tracker-App
# 187129562/fixbug/application not reviewed

### [Pivotal Tracker Link][tracker]


[tracker]: https://www.pivotaltracker.com/story/show/187129562

## What this PR does:


This pull request fixes the status the teacher can see after login to the account for two cases:
1. after being requested more info from an admin, and teacher login back 
-> teacher should see a yellow warning flash message about requested to update info. (change from not_reviewed to info_needed status)
2. after the teacher updates the info (with meaningful changes, i.e. not the same info as last time)
-> teacher should see a red alert flash message about not reviewed (change from info_needed to not_reviewed)

### Include screenshots, videos, etc.
#### Original Problem from previous semester
<img width="807" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/b3ea12a5-ede8-4748-9c34-e99a87389b7f">

#### Current effect after fix
##### Case1: Teacher logged in, but has not update info yet
<img width="1329" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/02941cb4-c106-4783-af27-01eeef84ee95">

##### Case2: Teacher logged in, with info updated
<img width="807" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/b3ea12a5-ede8-4748-9c34-e99a87389b7f">

#### High possibility of original cause of the problem
Teacher db schema has fields other than the info that can be submitted in the form:
<img width="664" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/25d8de53-44fe-45ea-be27-2803538300da">

The original code seems to trigger the `before_update` more often  due to the change of other unrelated fields, and thus cause the reset from `info_needed` to `not_reviewed`:
<img width="491" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/104db195-f589-46a0-83e6-d17b76317e8a">

#### Who authored this PR?

1. Perry (Jingchao) Zhong
2. Michael Tao


### How should this PR be tested?

* Is there a deploy we can view?
* What do the specs/features test?
1. Test expected behavior when the teacher get a request to update info, and finish the update
    - test the correct status
    - test the alert message and styling
    - test the update without meaningful change
    
* Are there edge cases to watch out for?

#### Are there any complications to deploying this?


No.

### Checklist:
